tfx_addons/sampler/executor.py

```diff
@@ -0,0 +1,259 @@
+      - sampled_examples: sampled examples, only for the given
+      splits as specified in splits. May also include copies of the
+      other non-sampled spits, as specified by null_classes.
+      exec_properties: A dict of execution properties, including:
+      - name: Optional unique name. Necessary if multiple components are
+      declared in the same pipeline.
+      - label: The name of the column containing class names to sample by.
+      - splits: A list containing splits to sample. Defaults to ['train'].
+      - copy_others: Determines whether we copy over the splits that aren't
+      sampled, or just exclude them from the output artifact. Defualts
+      to True.
+      - shards: The number of files that each sampled split should
+      contain. Default 0 is Beam's tfrecordio function's default.
+      - null_classes: A list determining which classes that we should
+      not sample. Defaults to None.
+    Returns:
+      None
+    """
+
+    self._log_startup(input_dict, output_dict, exec_properties)
+
+    label = exec_properties[spec.SAMPLER_LABEL_KEY]
+    sampling_strategy = exec_properties[spec.SAMPLER_SAMPLE_KEY]
+    splits = json_utils.loads(exec_properties[spec.SAMPLER_SPLIT_KEY])
+    copy_others = exec_properties[spec.SAMPLER_COPY_KEY]
+    shards = exec_properties[spec.SAMPLER_SHARDS_KEY]
+    null_classes = json_utils.loads(exec_properties[spec.SAMPLER_CLASSES_KEY])
+
+    input_artifact = artifact_utils.get_single_instance(
+        input_dict[spec.SAMPLER_INPUT_KEY])
+    output_artifact = artifact_utils.get_single_instance(
+        output_dict[spec.SAMPLER_OUTPUT_KEY])
+
+    if copy_others:
+      output_artifact.split_names = input_artifact.split_names
+    else:
+      output_artifact.split_names = artifact_utils.encode_split_names(splits)
+
+    # Fetch the input uri for each split
+    split_data = {}
+    for split in artifact_utils.decode_split_names(input_artifact.split_names):
+      uri = artifact_utils.get_split_uri([input_artifact], split)
+      split_data[split] = uri
+
+    for split, uri in split_data.items():
+      if split in splits:  # Undersampling split
+        output_dir = artifact_utils.get_split_uri([output_artifact], split)
+        split_dir = os.path.join(output_dir, f"Split-{split}")
+        with self._CreatePipeline(split_dir) as p:
+          sample(p, uri, label, shards, null_classes, split_dir,
+                 sampling_strategy)
+      elif copy_others:  # Copy the other split if copy_others is True
+        input_dir = uri
+        output_dir = artifact_utils.get_split_uri([output_artifact], split)
+        for filename in fileio.listdir(input_dir):
+          input_uri = os.path.join(input_dir, filename)
+          output_uri = os.path.join(output_dir, filename)
+          io_utils.copy_file(src=input_uri, dst=output_uri, overwrite=True)
+
+
+def generate_elements(x, label):
+  """Funciton that fetches the class label from a tf.Example and returns one
```

Comment:
"Function" . Also this is complicated enough I'd consider a real docstring. I don't know what `x` is, intuitively